browser/brave_cosmetic_resources_tab_helper.cc

```diff
@@ -0,0 +1,461 @@
+      cosmeticFilterConsiderNewSelectors_script +=
+            "window.cosmeticStyleSheet.insertRule(`" + styled_stylesheet +
+                "`, nextIndex);"
+            "window.allSelectorsToRules.set(`" +
+                styled_stylesheet + "`, nextIndex);"
+            "nextIndex++;";
+      cosmeticFilterConsiderNewSelectors_script +=
+            "if (!document.adoptedStyleSheets.includes("
+                  "window.cosmeticStyleSheet)){"
+               "document.adoptedStyleSheets = [window.cosmeticStyleSheet];"
+            "};";
+      cosmeticFilterConsiderNewSelectors_script +=
+          "})();";
+      auto* frame_host = content::RenderFrameHost::FromID(frame_id);
+      if (!frame_host)
+        return;
+      frame_host->ExecuteJavaScriptInIsolatedWorld(
+          base::UTF8ToUTF16(cosmeticFilterConsiderNewSelectors_script),
+          base::NullCallback(), ISOLATED_WORLD_ID_CHROME_INTERNAL);
+    }
+  }
+
+  if (!enabled_1st_party_cf_filtering_) {
+    auto* frame_host = content::RenderFrameHost::FromID(frame_id);
+    if (!frame_host)
+      return;
+    frame_host->ExecuteJavaScriptInIsolatedWorld(
+        base::UTF8ToUTF16(
+            *BraveCosmeticResourcesTabHelper::observing_script_),
+        base::NullCallback(), ISOLATED_WORLD_ID_CHROME_INTERNAL);
+  }
+}
+
+void BraveCosmeticResourcesTabHelper::GetHiddenClassIdSelectorsOnUI(
+    content::GlobalFrameRoutingId frame_id, const GURL& url,
+    std::unique_ptr<base::ListValue> selectors) {
+  if (!selectors || IsVettedSearchEngine(url.host())) {
+    return;
+  }
+  if (enabled_1st_party_cf_filtering_) {
+    // (sergz) We do the same in else section, we need to figure out do we
+    // need that if at all
+  } else {
+    std::string cosmeticFilterConsiderNewSelectors_script =
+        "(function() {"
+          "let nextIndex = window.cosmeticStyleSheet.rules.length;";
+    bool execute_script = false;
+    for (size_t i = 0; i < selectors->GetSize(); i++) {
+      base::ListValue* selectors_list = nullptr;
+      if (!selectors->GetList()[i].GetAsList(&selectors_list) ||
+          selectors_list->GetSize() == 0) {
+        continue;
+      }
+      for (size_t j = 0; j < selectors_list->GetSize(); j++) {
+        std::string rule = selectors_list->GetList()[i].GetString() +
+            "{display:none !important;}";
+        cosmeticFilterConsiderNewSelectors_script +=
+          "if (!window.allSelectorsToRules.has(`" +
+            selectors_list->GetList()[i].GetString() + "`)) {"
+            "window.cosmeticStyleSheet.insertRule(`" + rule +
+              "`,nextIndex);"
+            "window.allSelectorsToRules.set(`" +
+              selectors_list->GetList()[i].GetString() +
+              "`, nextIndex);"
+            "nextIndex++;"
+            "window.firstRunQueue.add(`" +
+                selectors_list->GetList()[i].GetString() + "`);"
+          "}";
+        execute_script = true;
+      }
```

Comment:
I imagine this ends up generating an awful lot of JS in `cosmeticFilterConsiderNewSelectors_script` - any reason not to generate a list and iterate over that with a loop in JS code?

Same applies to `CSSRulesRoutine`

Reply:
`frame_host->ExecuteJavaScriptInIsolatedWorld` accepts a formed JS to execute. Even if we have a function there to accept a list, we need to pass all that items there(the same forming of JS with that list and it's values and calling the function). I don't see how that could lead to a smaller JS string.

Reply:
If the average selector is `n` characters long, you're inserting `4n + ~200` characters _for each selector_, including the surrounding code.

Alternatively, building up a list like this would just require `n + 3` characters per selector:
```c++
cosmeticFilterConsiderNewSelectors_script += "const selectors = [";
for (size_t j = 0; j < selectors_list->GetSize(); j++) {
  cosmeticFilterConsiderNewSelectors_script += "`" + selectors_list->GetList()[i].GetString() + "`,"
}
cosmeticFilterConsiderNewSelectors_script += "];";
cosmeticFilterConsiderNewSelectors_script += "selectors.forEach(selector => {/* ... */})";
```

Last I checked there are generally ~800 selectors injected per page. With a rough estimate of `n = 10`, that translates to nearly a 20x difference in the size of the injected code.

Reply:
ok, you are right here, it will decrease it's size, will do that

Reply:
done! Good catch, number of JS chars we passed before was 328731, and it is 37105 after the optimization ~10x less.